csharp/ql/src/Likely Bugs/PossiblyIncorrectVariable.ql

```diff
@@ -0,0 +1,170 @@
+      this instanceof Property
+    )
+  }
+
+  Scope getScope() {
+    this.(LocalVariable).getVariableDeclExpr().getEnclosingStmt() = result
+          .(BlockStmt)
+          .getAChildStmt() or
+    this.(Parameter).getDeclaringElement() = result or
+    this.(Member).getDeclaringType() = result
+  }
+}
+
+class ComparisonOrAssignment extends Expr {
+  ComparisonOrAssignment() {
+    this instanceof AssignExpr or
+    this instanceof EqualityOperation or
+    this instanceof LocalVariableDeclAndInitExpr
+  }
+
+  Access getLeft() { result = this.getChild(0) }
+
+  Access getRight() { result = this.getChild(1) }
+}
+
+/**
+ * Holds if two comparisons or assignments are adjacent, either within an expression
+ * or in a sequence of assignments or declarations.
+ */
+predicate adjacent(ComparisonOrAssignment a1, ComparisonOrAssignment a2) {
+  exists(BinaryLogicalOperation expr | a1 = expr.getLeftOperand() |
+    a2 = expr.getRightOperand() or a2 = expr.getRightOperand().(LogicalAndExpr).getLeftOperand()
+  )
+  or
+  exists(BlockStmt block, ExprStmt s1, ExprStmt s2, int i |
+    s1 = block.getChildStmt(i) and s2 = block.getChildStmt(i + 1)
+  |
+    a1 = s1.getExpr() and a2 = s2.getExpr()
+  )
+  or
+  exists(BlockStmt block, LocalVariableDeclStmt s1, LocalVariableDeclStmt s2, int i |
+    s1 = block.getChildStmt(i) and s2 = block.getChildStmt(i + 1)
+  |
+    a1 = s1.getAVariableDeclExpr() and a2 = s2.getAVariableDeclExpr()
+  )
+  or
+  exists(LocalVariableDeclStmt decl, int i |
+    a1 = decl.getVariableDeclExpr(i) and a2 = decl.getVariableDeclExpr(i + 1)
+  )
+}
+
+class SameVariables extends StructuralComparisonConfiguration {
+  SameVariables() { this = "Same variables in operation" }
+
+  override predicate candidate(ControlFlowElement e1, ControlFlowElement e2) {
+    duplicatedAccess1(e1, e2, _, _)
+  }
+}
+
+/*
+ * Holds for sequences of expressions of the form
+ * ```
+ * same1 = different1;
+ * same2 = different2;
+ * ```
+ * where the target of `same1` and `same1` is equal.
+ */
+
+pragma[noopt]
+predicate duplicatedAccess1(Access same1, Access same2, Access different1, Access different2) {
+  exists(ComparisonOrAssignment op1, ComparisonOrAssignment op2 | adjacent(op1, op2) |
+    same1 = op1.getLeft() and
+    different1 = op1.getRight() and
+    same2 = op2.getLeft() and
+    different2 = op2.getRight()
+    or
+    same1 = op1.getRight() and
+    different1 = op1.getLeft() and
+    same2 = op2.getRight() and
+    different2 = op2.getLeft()
+  ) and
+  same1.getTarget() = same2.getTarget()
+}
+
+predicate duplicatedAccess2(Access same1, Access same2, Access different1, Access different2) {
+  duplicatedAccess1(same1, same2, different1, different2) and
+  any(SameVariables same).same(same1, same2) and
+  exists(Declaration target | target = same1.getTarget() |
+    not target.(Modifiable).isStatic() and
+    not target instanceof ImplicitAccessorParameter and
+    not target instanceof EnumConstant
```

Comment:
How about moving these restrictions into `ScopedDeclaration` and replace `exists(Declaration target` with `exists(ScopedDeclaration target`?